src/language-services/formatter.ts

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as PQF from "@microsoft/powerquery-formatter";
+import * as PQP from "@microsoft/powerquery-parser";
+
+import { FormattingOptions, Range, TextDocument, TextEdit } from "vscode-languageserver-types";
+
+export function tryFormatconnection(
+    document: TextDocument,
+    formattingOptions: FormattingOptions,
+    localizationTemplates: PQP.ILocalizationTemplates,
+): PQP.Result<TextEdit[], string> {
+    let indentationLiteral: PQF.IndentationLiteral;
+    if (formattingOptions.insertSpaces) {
+        indentationLiteral = PQF.IndentationLiteral.SpaceX4;
+    } else {
+        indentationLiteral = PQF.IndentationLiteral.Tab;
+    }
+
+    const pqfFormatSettings: PQF.FormatSettings = {
+        ...PQP.DefaultSettings,
+        localizationTemplates,
+        indentationLiteral,
+        newlineLiteral: PQF.NewlineLiteral.Windows,
+    };
+    const triedFormat: PQF.TriedFormat = PQF.tryFormat(pqfFormatSettings, document.getText());
+
+    if (PQP.ResultUtils.isOk(triedFormat)) {
+        return PQP.ResultUtils.okFactory([TextEdit.replace(fullDocumentRange(document), triedFormat.value)]);
+    } else {
+        let message: string;
+        if (PQF.FormatError.isTFormatError(triedFormat.error)) {
+            message = triedFormat.error.innerError.message;
+        } else {
+            message = "An unknown error occured during formatting.";
```

Comment:
Do we have other cases where we have to localize strings in language-services?

Reply:
Discussed in Teams:

We're going to either throw an error message from the parser (already localized) or throw undefined and let VSCode handle it from there.